plugin/loop/loop.go

```diff
@@ -0,0 +1,94 @@
+package loop
+
+import (
+	"context"
+	"os"
+	"sync"
+
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+var log = clog.NewWithPlugin("loop")
+
+// Loop is a plugin that implements loop detection by sending a "random" query.
+type Loop struct {
+	Next plugin.Handler
+
+	zone  string
+	qname string
+
+	sync.RWMutex
+	i   int
+	off bool
+}
+
+// New returns a new initialized Loop.
+func New(zone string) *Loop { return &Loop{zone: zone, qname: qname(zone)} }
+
+// ServeDNS implements the plugin.Handler interface.
+func (l *Loop) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	if r.Question[0].Qtype != dns.TypeHINFO {
+		return plugin.NextOrFailure(l.Name(), l.Next, ctx, w, r)
+	}
+	if l.disabled() {
+		return plugin.NextOrFailure(l.Name(), l.Next, ctx, w, r)
+	}
+
+	state := request.Request{W: w, Req: r}
+
+	zone := plugin.Zones([]string{l.zone}).Matches(state.Name())
+	if zone == "" {
+		return plugin.NextOrFailure(l.Name(), l.Next, ctx, w, r)
+	}
+
+	if state.Name() == l.qname {
+		l.inc()
+	}
+
+	if l.seen() > 2 {
+		// TODO(miek): add log.Fatal(f) and use here.
```

Comment:
I understand we create only one Loop struct. and we send several queries (every 2 sec).
so, this l.seen > 2 will be true at the third msg we send .. no ?

Should we have here a count for each msg.id we create for the test ?